### PR TITLE
feat(swarm): report outcome of handling `SwarmEvent`

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

Previously, a user wouldn't know whether passing a `SwarmEvent` to `ListenAddresses` or `ExternalAddresses` changed the state. We now return a boolean where `true` indicates that we handled the event **and** changed state as a result.

The API is inspired by `HashSet::insert` and the like.

## Notes & open questions

Extracted out of #3651.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: swarm/CHANGELOG.md [hunk 1]
```diff
@@ -3,7 +3,11 @@
 - Raise MSRV to 1.65.
   See [PR 3715].
 
+- Return a bool from `ExternalAddresses::on_swarm_event` and `ListenAddresses::on_swarm_event` indicating whether any state was changed.
+  See [PR XXXX].
+
 [PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
  See [PR 3865].

[PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
[PR 3865]: https://github.com/libp2p/rust-libp2p/pull/3865
```

File: swarm/src/behaviour/external_addresses.rs [hunk 2]
```diff
@@ -32,21 +32,67 @@ impl ExternalAddresses {
...
+    fn new_external_addr() -> FromSwarm<'static, dummy::ConnectionHandler> {
+        FromSwarm::NewExternalAddr(NewExternalAddr { addr: &MEMORY_ADDR })
+    }
+
+    fn expired_external_addr() -> FromSwarm<'static, dummy::ConnectionHandler> {
+        FromSwarm::ExpiredExternalAddr(ExpiredExternalAddr { addr: &MEMORY_ADDR })
     }
+
+    static MEMORY_ADDR: Lazy<Multiaddr> =
+        Lazy::new(|| Multiaddr::empty().with(Protocol::Memory(1000)));
```
Comment (mxinden) [inline swarm/src/behaviour/external_addresses.rs:97]: What is the value of a `static` over a `const` here? Why does it warrant introducing `once_cell` as a dependency?

**Review by mxinden — DISMISSED** [review]
Looks good to me. Though I would prefer not using `once_cell::Lazy` here for the sake of simplicity.

File: swarm/src/behaviour/external_addresses.rs (hunk 2 above)
Comment (thomaseizinger) [reply swarm/src/behaviour/external_addresses.rs:97]: `Multiaddr::empty` and `Multiaddr::with` are unfortunately not `const`.

Comment (mxinden) [reply swarm/src/behaviour/external_addresses.rs:97]: Ah, right. I missed that.

I am hesitant to introduce a dependency just for the sake of more succinct test code. That said, `once_cell` seems to be everywhere, thus not really an additional dependency.

**Review by mxinden — APPROVED** [review]

File: swarm/src/behaviour/external_addresses.rs (hunk 2 above)
Comment (thomaseizinger) [reply swarm/src/behaviour/external_addresses.rs:97]: It is on the pathway to being stabilized and part of std: https://github.com/rust-lang/rust/pull/105587

**Pull request merged**